src/base/utils/foreignapps.cpp

```diff
@@ -254,20 +258,43 @@ bool Utils::ForeignApps::PythonInfo::isSupportedVersion() const
 PythonInfo Utils::ForeignApps::pythonInfo()
 {
     static PythonInfo pyInfo;
-    if (!pyInfo.isValid())
+
+    const QString preferredPythonPath = Preferences::instance()->getPythonExecutablePath().toString();
+    if (pyInfo.isValid() && (preferredPythonPath == pyInfo.executableName))
+        return pyInfo;
+
+    if (!preferredPythonPath.isEmpty())
     {
-        if (testPythonInstallation(u"python3"_s, pyInfo))
+        if (testPythonInstallation(preferredPythonPath, pyInfo))
             return pyInfo;
+        LogMsg(QCoreApplication::translate("Utils::ForeignApps", "Failed to find Python executable. Path: \"%1\".")
+            .arg(preferredPythonPath), Log::WARNING);
+    }
+    else
+    {
+        // auto detect only when there are no preferred python path
 
-        if (testPythonInstallation(u"python"_s, pyInfo))
-            return pyInfo;
+        if (!pyInfo.isValid())
```

Comment:
This check is redundant. It can refer to valid, but another python installation.

IIRC, there is no way currently to avoid re-checking python installation in all the possible cases unless you store previously used custom path, but I strongly disapprove of a utility function doing this (it already requires a redesign in order not to be dependent on anything other than its parameters, otherwise it is something more than some kind of utility helper).
But I really expect that it returns "info" that matches current "custom python path" value.

Reply:
>But I really expect that it returns "info" that matches current "custom python path" value.

The auto detect is slow and this function will be called multiple times.
I can already guess users won't be satisfy with you and I won't like to do it in my PR.

>This check is redundant. It can refer to valid, but another python installation.

And pardon, I currently can't afford the time to redesign and make the experience better. Hopefully the current form will be enough for those who demands it.

Reply:
> The auto detect is slow and this function will be called multiple times.
> I can already guess users won't be satisfied with you and I won't like to do it in my PR.

Well, let's leave it to "restart required" caveat for now.